src/pykeen/training/callbacks.py

```diff
@@ -374,9 +384,117 @@ def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # n
             self.last_best_epoch = epoch
 
 
-callback_resolver: ClassResolver[TrainingCallback] = ClassResolver.from_subclasses(
-    base=TrainingCallback,
-)
+class OptimizerTrainingCallback(TrainingCallback):
+    """Use optimizer to update parameters."""
+
+    def __init__(self, step: bool = True):
+        """Initialize the callback.
+
+        :param step:
+            whether to apply the parameter update step
+        """
+        super().__init__()
+        self.step = step
+
+    # docstr-coverage: inherited
+    def pre_batch(self, **kwargs: Any) -> None:  # noqa: D102
+        # Recall that torch *accumulates* gradients. Before passing in a
+        # new instance, you need to zero out the gradients from the old instance
+        self.optimizer.zero_grad(set_to_none=True)
+
+    # docstr-coverage: inherited
+    def post_batch(self, epoch: int, batch, **kwargs: Any) -> None:  # noqa: D102
+        # when called by batch_size_search(), the parameter update should not be applied.
+        if self.step:
+            # TODO: callback pre_step
+            # update parameters according to optimizer
+            self.optimizer.step()
+
+        # After changing applying the gradients to the embeddings, the model is notified that the forward
+        # constraints are no longer applied
+        self.model.post_parameter_update()
+
+
+class LearningRateSchedulerTrainingCallback(TrainingCallback):
+    """Update learning rate scheduler."""
+
+    # docstr-coverage: inherited
+    def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # noqa: D102
+        if self.training_loop.lr_scheduler is not None:
+            self.training_loop.lr_scheduler.step(epoch=epoch)
+
+
+def _hasher(kwargs: Mapping[str, Any]) -> int:
+    # do not share optimal parameters across different training loops
+    return id(kwargs["training_loop"])
+
+
+@maximize_memory_utilization(parameter_name=("batch_size", "slice_size"), hasher=_hasher)
+@torch.inference_mode()
+def _validation_loss_amo_wrapper(
+    training_loop: "TrainingLoop",
+    triples_factory: CoreTriplesFactory,
+    batch_size: int,
+    slice_size: int,
+    label_smoothing: float,
+    **kwargs,
+) -> float:
+    """Calculate validation loss with automatic batch size optimization."""
+    return training_loop._train_epoch(
+        batches=training_loop._create_training_data_loader(
+            triples_factory=triples_factory, batch_size=batch_size, drop_last=False, **kwargs
+        ),
+        label_smoothing=label_smoothing,
+        # no callbacks, no epoch
+        callbacks=MultiTrainingCallback(),
+        epoch=-1,
```

Comment:
can this negative number mess some unsuspecting callbacks up?

Reply:
no, since we combine it with an empty callback (`MultiTrainingCallback`)